src/gtkspellcheck/spellcheck.py

```diff
@@ -807,3 +807,71 @@ def _continue_batched_recheck(self, start_mark):
             end.forward_char()
             start_mark = self._buffer.create_mark(None, end)
             GLib.idle_add(self._continue_batched_recheck, start_mark)
+
+    @staticmethod
+    def _between_middle_and_end_of_word(loc):
+        if loc.starts_word():
+            if loc.is_start():
+                return False
+
+            # Determine if actually in word after extra chars
+            loc = loc.copy()
+            loc.backward_char()
+            if not SpellChecker._is_extra_word_char(loc):
+                return False
+
+            if loc.is_start():
+                return False
+            loc.backward_char()
+            return loc.inside_word()
+        else:
+            return loc.inside_word() or loc.ends_word()
+
+    @staticmethod
+    def _is_extra_word_char(loc):
+        ch = loc.get_char()
+
+        if ch == " " or ch == "\n" or ch == "\t" or ch == "\r":
+            return False
+
+        if ch == "'":
+            return True
+
+        # Language extra chararacters should also be processed once Enchant's
+        # enchant_dict_get_extra_word_characters is exposed in PyEnchant
+
+        return False
+
+    @staticmethod
+    def _iter_forward_word_end(loc):
+        tmp = loc.copy()
+        if loc.forward_word_end():
+            tmp = loc.copy()
+
+            if SpellChecker._is_extra_word_char(tmp):
+                if SpellChecker._iter_forward_word_end(tmp):
+                    loc.assign(tmp)
+
+            return True
+
+        if loc.is_end() and loc.ends_word() and not tmp.equal(loc):
+            return True
+
+        return False
```

Comment:
This also seems quite complex.

Wouldn't the following work as well?

```python
loc.forward_word_end()
while SpellChecker._is_extra_word_char(loc):
   loc.forward_word_end()
```

Reply:
This suggestion resulted in a loop when the extra char was at the end of the buffer. I've reverted to the original for the moment.

Reply:
I still think that the recursive solution is too complex.

Can't we just add an additional condition for aborting the loop when being at the end of the buffer? Or, adapt the implementation of `is_extra_word_char` to return `False` for the end of the buffer?

Reply:
I've pushed something up that's simpler. But... I wonder how this gets tested in a useful way? The more complex implementation, although being clunky, has in theory had a good deal more of a working over by somebody a great deal more familiar with the `Gtk.TextIter` API etc.

This is primarily a question from somebody who hasn't the time to take this through the more thorough exercise that I'd prefer. However "it works for me for the few test cases I've tried" :slightly_smiling_face:

Reply:
Perhaps having the extra chars behaviour disabled by default would alleviate some of my doubts. It would be a flag, different to just an empty chars set, so that it doesn't run through the different logic.

Reply:
I appreciate your concerns, we should indeed make sure that this works as expected.

After reading the documentation, I think that the following is actually the correct solution:
```python
while loc.forward_word_end() and is_extra_word_char(loc):
    pass
```
The method `forward_word_end` will return `true` if and only if `loc` has been moved. Thus, in the worst case, it will consume the whole buffer or abort too early. However, this loop should never diverge (which would be really bad because it would just hang the program indefinitely). This will abort if for whatever reason (end of buffer or who knows what) `loc` does no longer move.

Reply:
Hey, thanks for nudging this along. 

The issue with this approach is that if `loc` points at an extra word char the iterator won't get moved in `forward_word_end` which will then return false and we don't move through the extra char.

Reply:
I guess, I still do not fully understand how this function works.

In the initial implementation, there is no call to `forward_char`. I wonder how the initial implementation accounts for `forward_word_end` not moving in this case?

Sorry for all the questions, I just hesitate to merge code which I do not quite understand. 

What do you think about the following implementation:

```python
def forward_word_end(self, loc):
    loc.forward_word_end()
    while self.is_extra_word_char(loc):
        loc.forward_char()
        if loc.is_end() or not loc.forward_word_end():
            break
```
Again, the worst thing that can happen with this implementation is consuming the whole buffer or aborting too early. I think that this is at least something we want to be absolutely sure of.

Reply:
100% fine with all the questions, sorry the process here is all a bit out of whack with me having copied/ported logic from elsewhere, planned to refine that later and then lost all my time. Kind of rubbish, sorry! :facepalm:

Right, so I think the initial implementation uses `forward_word_char` to move to the section after the special char, which it sees as another word. 

>If iter is currently on a word end, moves forward to the next one after that

[forward_word_end](https://docs.gtk.org/gtk4/method.TextIter.forward_word_end.html)

One of the downsides of that approach is that eg. while attempting to write `isn't` at the end of the buffer `isn` will be marked as an incorrect word when one has input `isn'`, which is maybe fine if not perfect (and fwiw appears to be what Firefox does as I'm typing this).

Your suggestion (and what's currently in the repo) avoid that.

However all three versions (the initial, current and proposed) seem to have an issue: if a misspelled word ends in an extra word char, appending another word afterwards appears to result in the misspelled word being marked valid. Text Editor doesn't exhibit this behaviour. Potentially not related to `forward_word_end`. Will attempt to grab time in the next couple of days to look into it.

Reply:
Thanks for the clarifications.

I think this issue can be addressed by checking for `inside_word` in the `if` condition:

```python
def forward_word_end(self, loc):
    loc.forward_word_end()
    while self.is_extra_word_char(loc):
        loc.forward_char()
        if loc.is_end() or not loc.inside_word() or not loc.forward_word_end():
            break
```

This may, however, trigger the same issue in case there are multiple extra word chars. Another solution could be to move forward char by char as long as `is_extra_word_char` and then, if the current char is not an extra word char, check whether we are still inside a word – if this is the case, then consume the word and repeat.

Reply:
Thanks! I've pushed some changes based on this, aiming for something that would accommodate for multiple extra chars in a row.